crypto/sm2/sm2_pmeth.c

```diff
@@ -0,0 +1,306 @@
+        return 0;
+    }
+
+    ctx->data = dctx;
+    return 1;
+}
+
+static int pkey_sm2_copy(EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src)
+{
+    SM2_PKEY_CTX *dctx, *sctx;
+
+    if (!pkey_sm2_init(dst))
+        return 0;
+    sctx = src->data;
+    dctx = dst->data;
+    if (sctx->gen_group != NULL) {
+        dctx->gen_group = EC_GROUP_dup(sctx->gen_group);
+        if (dctx->gen_group != NULL)
+            return 0;
+    }
+    dctx->md = sctx->md;
+
+    return 1;
+}
+
+static void pkey_sm2_cleanup(EVP_PKEY_CTX *ctx)
+{
+    SM2_PKEY_CTX *dctx = ctx->data;
+
+    if (dctx != NULL) {
+        EC_GROUP_free(dctx->gen_group);
+        OPENSSL_free(dctx);
+    }
+}
+
+static int pkey_sm2_sign(EVP_PKEY_CTX *ctx, unsigned char *sig, size_t *siglen,
+                        const unsigned char *tbs, size_t tbslen)
+{
+    int ret;
+    unsigned int sltmp;
+    SM2_PKEY_CTX *dctx = ctx->data;
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+
+    const int type = (dctx->md == NULL) ? NID_sm3 : EVP_MD_type(dctx->md);
+
+    if (sig == NULL) {
+        *siglen = ECDSA_size(ec);
+        return 1;
+    }
+
+    if (*siglen < (size_t)ECDSA_size(ec)) {
+        SM2err(SM2_F_PKEY_SM2_SIGN, SM2_R_BUFFER_TOO_SMALL);
+        return 0;
+    }
+
+    ret = sm2_sign(type, tbs, tbslen, sig, &sltmp, ec);
+
+    if (ret <= 0)
+        return ret;
+    *siglen = (size_t)sltmp;
+    return 1;
+}
+
+static int pkey_sm2_verify(EVP_PKEY_CTX *ctx,
+                           const unsigned char *sig, size_t siglen,
+                           const unsigned char *tbs, size_t tbslen)
+{
+    SM2_PKEY_CTX *dctx = ctx->data;
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+    const int md_type = (dctx->md == NULL) ? NID_sm3 : EVP_MD_type(dctx->md);
+
+    return sm2_verify(md_type, tbs, tbslen, sig, siglen, ec);
+}
+
+static int pkey_sm2_encrypt(EVP_PKEY_CTX *ctx,
+                              unsigned char *out, size_t *outlen,
+                              const unsigned char *in, size_t inlen)
+{
+    int ret;
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+    SM2_PKEY_CTX *dctx = ctx->data;
+    const int md_type = (dctx->md == NULL) ? NID_sm3 : EVP_MD_type(dctx->md);
+
+    if (out == NULL) {
+        if (!sm2_ciphertext_size(ec, EVP_get_digestbynid(md_type), inlen,
+                                 outlen))
+            ret = -1;
+        else
+            ret = 1;
+    } else {
+        ret = sm2_encrypt(ec, EVP_get_digestbynid(md_type),
+                          in, inlen, out, outlen);
+    }
+
+    return ret;
+}
+
+static int pkey_sm2_decrypt(EVP_PKEY_CTX *ctx,
+                              unsigned char *out, size_t *outlen,
+                              const unsigned char *in, size_t inlen)
+{
+    int ret = -1;
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+    SM2_PKEY_CTX *dctx = ctx->data;
+    const int md_type = (dctx->md == NULL) ? NID_sm3 : EVP_MD_type(dctx->md);
+
+    if (out == NULL) {
+        if (!sm2_plaintext_size(ec, EVP_get_digestbynid(md_type), inlen,
```

Comment:
I don't quite follow why does EVP_get_digestbynid have to be invovled here? As simplest case wouldn't `dctx->md ? dctx->md : EVP_sm3()` do the job? If one has to use EVP_get_digestbynid(NID_sm3), then `dctx->md ? dctx->md : EVP_get_digestbynid(NID_sm3). Either way this kind of calls for question if it would be appropriate to check for EVP_get_digestbynid return value. Well, no-sm3 implies no-sm2, so one can argue that lookup for NID_sm3 is bound to succeed [as well as availability of EVPsm3()].